### ISSUE TITLE
Metadata containing date is parsed differently when in expanded entity

### ISSUE THREAD

**Issue by slimCODE** [stripe/stripe-dotnet#1178]
Given a metadata on a customer containing the output of `DateTimeOffset.ToString("o")`, as in "2018-04-30T19:03:07.6062218+00:00", calling `StripeSubscriptionService.ListAsync` with `ExpandCustomer` set to true returns a `Customer.Metadata[key]` containing a different date (local and parsed), like "04/30/2018 15:03:07".

Sample code:

```c#
    // Given each customer has a "trialStartedDate" metadata
    var subService = new StripeSubscriptionService() { ExpandCustomer = true };
    var custService = new StripeCustomerService();

    var subscriptions = await subService.ListAsync();
    var firstSubscription = subscriptions.FirstOrDefault();
    var customer = await custService.GetAsync(firstSubscription.CustomerId);

    Console.WriteLine($"{firstSubscription.Customer.Metadata["trialStartedDate"]}");
    Console.WriteLine($"{customer.Metadata["trialStartedDate"]}");

```

The output is:

> 04/30/2018 15:03:07

> 2018-04-30T19:03:07.6062218+00:00

Tested with version 13.3.1 and 15.8.0.
Tested with a .NET Core app on Windows and Linux.
Confirmed the JSON payload is the same, so the conversion occurs in the SDK.

**Comment by fred-stripe** [comment]
I'm pretty sure this is due to automatic ISO date detection from JSON.net:

https://github.com/JamesNK/Newtonsoft.Json/blob/4c4d870816ffddcc98803b1ad64b345a2c9c8dfb/Src/Newtonsoft.Json/JsonSerializer.cs#L450-L458
```c#
        /// Gets or sets how date formatted strings, e.g. <c>"\/Date(1198908717056)\/"</c> and <c>"2012-03-21T05:40Z"</c>, are parsed when reading JSON.
        /// The default value is <see cref="Json.DateParseHandling.DateTime" />.
```

By default this will turn any ISO8601-looking string in a JSON object into a `DateTime`:

https://github.com/JamesNK/Newtonsoft.Json/blob/78e899fd50b82ff3e16f73c8a50772bf21157599/Src/Newtonsoft.Json.Tests/JsonTextReaderTests/ParseTests.cs#L445-L454

It looks like Stripe.net doesn't adjust the JSON serializer settings and so is getting the default behavior here. While unexpected this seems like it might be useful since you use ISO8601 dates in your metadata—what would you prefer to have happen here?

**Comment by slimCODE** [comment]
I expect the same data from `firstSubscription.Customer.Metadata["trialStartedDate"]` and `customer.Metadata["trialStartedDate"]`. This automatic conversion should not kick in "sometimes"(*). It makes code accessing a `StripeCustomer`'s Metadata fail when that entity comes from one place, and not the other.

(*): Which is really "when the entity is an expanded child entity of another".

**Comment by fred-stripe** [comment]
Hm, yeah, agreed this is really strange. It looks to be the same code path in Stripe.net and JSON.net so I'm not sure what's happening here. Will investigate further!

**Comment by slimCODE** [comment]
I added this to the code above:

```
    var singleSubscription = await subService.GetAsync(firstSubscription.Id);
    Console.WriteLine($"{singleSubscription.Customer.Metadata["trialStartedDate"]}");
```

And it displays "04/30/2018 15:03:07" just like `firstSubscription`, so it's not about `StripeList` handling or a difference between `ListAsync` vs `GetAsync`, but really child entities.

**Comment by jaymedavis** [comment]
[This may help you](https://stackoverflow.com/questions/2423377/what-is-the-invariant-culture). Stripe.net converts date times to UTC, but only those specifically using a date time type. CultureInfo.InvariantCulture is probably what you need.

**Comment by slimCODE** [comment]
@jaymedavis I don't see how it relates to the issue. No matter what is the current culture, `Metadata[key]` on a `StripeCustomer` should always return the same `string`, no matter if that customer comes from `StripeCustomerService.GetAsync(...)` or the `Customer` property of a `StripeSubscription` from `StripeSubscriptionService.GetAsync(...)`.

**Comment by jaymedavis** [comment]
Ah, I see. If it helps, the expandable properties use [StringOrObject](https://github.com/stripe/stripe-dotnet/blob/master/src/Stripe.net/Infrastructure/StringOrObject.cs) to convert. Perhaps something in the casting to JToken (json.net) is causing an issue?

> Confirmed the JSON payload is the same, so the conversion occurs in the SDK.

How did you confirm this? Did you view the StripeResponse property or inspect the response at a deeper level?

**Comment by slimCODE** [comment]
With Fiddler. ;)

**Comment by remi-stripe** [comment]
Hey everyone! Sorry for the delays on that one, I think it fell through the cracks. I picked it up this morning and was able to reproduce the issue on 13.3.1. On the latest version though (21.2.0) the format is similar on both sides though there's still a timezone difference as I get `2018-04-30T19:03:07.6062218+00:00` on the customer object and `2018-04-30T21:03:07.6062218+02:00` when fetching or listing subscriptions.

This should solve the original issue but we're investigating how to ensure the date stays identical no matter which approach is taken.

**Comment by ob-stripe** [comment]
This should now be completely fixed as of version 22.0.0. The root cause of this issue was Newtonsoft.Json's strange default behavior when it comes to date-like strings (cf. https://github.com/JamesNK/Newtonsoft.Json/issues/862). The library now uses the custom setting `DateParseHandling.None` to disable this behavior in Newtonsoft.Json.
